flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/MiniDispatcher.java

```diff
@@ -129,10 +129,12 @@ public MiniDispatcher(
     }
 
     @Override
-    protected CleanupJobState jobReachedTerminalState(ExecutionGraphInfo executionGraphInfo) {
+    protected CompletableFuture<CleanupJobState> jobReachedTerminalState(
+            ExecutionGraphInfo executionGraphInfo) {
         final ArchivedExecutionGraph archivedExecutionGraph =
                 executionGraphInfo.getArchivedExecutionGraph();
-        final CleanupJobState cleanupHAState = super.jobReachedTerminalState(executionGraphInfo);
+        final CompletableFuture<CleanupJobState> cleanupHAState =
```

Comment:
we should wait for the `jobReachedTerminalState` to complete before completing the shutdownFuture

Reply:
It's more a question about whether we want the HistoryServer archiving to have happened before shutting down the cluster. We came up with the conclusion that, if a user configures the HistoryServer, he/she might have the desire to have the result archived. In that case, we want to wait for the archiving to happen (keep in mind, the HistoryServer is no-op by default; i.e. the user has to explicitly configure Flink to use it).

I don't understand your comment about `Originally, the shutdownFuture is completed without waiting for completion of jobReachedTerminalState`. May you elaborate a bit more on that. As far as I can see, it did wait for the `jobReachedTerminalState` to complete (because it was synchronously called). It didn't wait for the archiving to finish, though, which is what we want to change as far as I understand. Or am I missing something here? 🤔